src/passes/OptimizeInstructions.cpp

```diff
@@ -1513,10 +1513,74 @@ struct OptimizeInstructions
     return getDroppedChildrenAndAppend(curr, result);
   }
 
-  bool trapOnNull(Expression* curr, Expression* ref) {
+  Expression* getResultOfFirst(Expression* first, Expression* second) {
+    return wasm::getResultOfFirst(
+      first, second, getFunction(), getModule(), getPassOptions());
+  }
+
+  // Optimize an instruction and the reference it operates on, under the
+  // assumption that if the reference is a null then we will trap. Returns true
+  // if we replaced the expression with something simpler. Returns false if we
+  // found nothing to optimize, or if we just modified or replaced the ref (but
+  // not the expression itself).
+  bool trapOnNull(Expression* curr, Expression*& ref) {
+    Builder builder(*getModule());
+
+    if (getPassOptions().trapsNeverHappen) {
+      // We can ignore the possibility of the reference being an input, so
+      //
+      //    (if
+      //      (condition)
+      //      (null)
+      //      (other))
+      // =>
+      //    (drop
+      //      (condition))
+      //    (other)
+      //
+      // That is, we will by assumption not read from the null, so remove that
+      // arm.
+      //
+      // TODO We could recurse here.
+      // TODO We could do similar things for casts (rule out an impossible arm).
+      if (auto* iff = ref->dynCast<If>()) {
+        if (iff->ifFalse) {
+          if (iff->ifTrue->type.isNull()) {
+            ref = builder.makeBlock({builder.makeDrop(iff->condition),
+                                     builder.makeDrop(iff->ifTrue),
```

Comment:
The `ifTrue` block should not be included because we're assuming it would never be executed at runtime. This would be different for a `select`.

Reply:
Good catch... I missed it in copy-paste, and the fuzzer somehow missed it too... fixed.